src/ORM/Connect/PDOQuery.php

```diff
@@ -22,16 +22,52 @@ class PDOQuery extends Query
      * Hook the result-set given into a Query class, suitable for use by SilverStripe.
      * @param PDOStatement $statement The internal PDOStatement containing the results
      */
-    public function __construct(PDOStatement $statement)
+    public function __construct(PDOStatement $statement, PDOConnector $conn)
     {
         $this->statement = $statement;
         // Since no more than one PDOStatement for any one connection can be safely
         // traversed, each statement simply requests all rows at once for safety.
         // This could be re-engineered to call fetchAll on an as-needed basis
-        $this->results = $statement->fetchAll(PDO::FETCH_ASSOC);
+
+        // Special case for Postgres
+        if ($conn->getDriver() == 'pgsql') {
+            $this->results = $this->fetchAllPgsql($statement);
+        } else {
+            $this->results = $statement->fetchAll(PDO::FETCH_ASSOC);
+        }
         $statement->closeCursor();
     }
 
+    /**
+     * Fetch a record form the statementw ith its type data corrected
+     * Necessary to fix float data retrieved from PGSQL
+     */
+    protected function fetchAllPgsql($statement)
+    {
+        $columnCount = $statement->columnCount();
+        $columnMeta = [];
+        for ($i = 0; $i<$columnCount; $i++) {
+            $columnMeta[$i] = $statement->getColumnMeta($i);
+        }
+
+        // Re-map fetched data using columnMeta
+        return array_map(
+            function ($rowArray) use ($columnMeta) {
+                $row = [];
+                foreach ($columnMeta as $i => $meta) {
+                    // Coerce floats from string to float
+                    // PDO PostgreSQL fails to do this
+                    if (isset($meta['native_type']) && preg_match('/^float/', $meta['native_type'])) {
```

Comment:
Given this will be called on an unknown number of rows for each query, could/should we replace this with `strpos($meta['native_type'], 'float') === 0`?

Reply:
I'm pretty sure that preg_match is the fastest match. But we could also potentially get a list of all the options.

Reply:
I stand corrected:
```
php test-preg.php 
preg... 3.5830228328705
strpos... 1.2603948116302
in_array... 1.4446079730988
```

```php
<?php

define('LOTS', 10000000);

profile('preg', function() {
  for ($i=0; $i<LOTS; $i++) {
    preg_match('/^float/', 'float8');
    preg_match('/^float/', 'float16');
    preg_match('/^float/', 'varchar');
    preg_match('/^float/', 'int');
  }
});

profile('strpos', function() {
  for ($i=0; $i<LOTS; $i++) {
    strpos('float8', 'float');
    strpos('float16', 'float');
    strpos('varchar', 'float');
    strpos('int', 'float');
  }
});

profile('in_array', function() {
    $array = ['float8', 'float16', 'float4'];
  for ($i=0; $i<LOTS; $i++) {
    in_array('float8', $array);
    in_array('float16', $array);
    in_array('varchar', $array);
    in_array('int', $array);
  }
});

function profile($label, $callback) {
    echo $label . "... ";
    $t1 = microtime(true);
    $callback();
    $t2 = microtime(true);
    echo ($t2-$t1) . "\n";
}
```

Reply:
Still pretty close; i'm going with === 0 as it reduces the risk that i'm missing an entry in my array ;)

```
preg... 4.3494780063629
strpos... 1.7755749225616
in_array... 1.6038949489594
```